lib/api/normalize.js

```diff
@@ -0,0 +1,307 @@
+     */
+    preserveAtomicBlocks(editorState, entityTypes) {
+        const content = editorState.getCurrentContent();
+        const blockMap = content.getBlockMap();
+
+        const perservedBlocks = blockMap
+            .filter(block => {
+                const entityKey = block.getEntityAt(0);
+
+                return (
+                    entityKey &&
+                    entityTypes.includes(content.getEntity(entityKey).getType())
+                );
+            })
+            .map(block => block.set('type', BLOCK_TYPE.ATOMIC));
+
+        if (perservedBlocks.size !== 0) {
+            return EditorState.set(editorState, {
+                currentContent: content.merge({
+                    blockMap: blockMap.merge(perservedBlocks),
+                }),
+            });
+        }
+
+        return editorState;
+    },
+
+    /**
+     * Resets the depth of all the content to at most maxListNesting.
+     */
+    resetBlockDepth(editorState, maxListNesting) {
+        const content = editorState.getCurrentContent();
+        const blockMap = content.getBlockMap();
+
+        const changedBlocks = blockMap
+            .filter(block => block.getDepth() > maxListNesting)
+            .map(block => block.set('depth', maxListNesting));
+
+        if (changedBlocks.size !== 0) {
+            return EditorState.set(editorState, {
+                currentContent: content.merge({
+                    blockMap: blockMap.merge(changedBlocks),
+                }),
+            });
+        }
+
+        return editorState;
+    },
+
+    /**
+     * Resets all blocks that use unavailable types to unstyled.
+     */
+    resetBlockType(editorState, enabledTypes) {
+        const content = editorState.getCurrentContent();
+        const blockMap = content.getBlockMap();
+
+        const changedBlocks = blockMap
+            .filter(block => !enabledTypes.includes(block.getType()))
+            .map(block => block.set('type', BLOCK_TYPE.UNSTYLED));
+
+        if (changedBlocks.size !== 0) {
+            return EditorState.set(editorState, {
+                currentContent: content.merge({
+                    blockMap: blockMap.merge(changedBlocks),
+                }),
+            });
+        }
+
+        return editorState;
+    },
+
+    /**
+     * Removes all styles that use unavailable types.
+     */
+    filterInlineStyle(editorState, enabledTypes) {
+        const content = editorState.getCurrentContent();
+        const blockMap = content.getBlockMap();
+
+        const blocks = blockMap.map(block => {
+            let altered = false;
+
+            const chars = block.getCharacterList().map(char => {
+                let newChar = char;
+
+                char
+                    .getStyle()
+                    .filter(type => !enabledTypes.includes(type))
+                    .forEach(type => {
+                        altered = true;
+                        newChar = CharacterMetadata.removeStyle(newChar, type);
```

Comment:
Reducing over the array to create a char without any styles. There may be a simpler way, but I doubt there would be chars with more than say 10 inline styles in a worst case scenario. 2-3 is already a lot for real-world content.